File: android_app_java/app/src/main/java/com/example/android_app_java/DashboardActivity.java
package com.example.android_app_java;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.os.Bundle;
import android.content.Intent;
import android.os.CountDownTimer;
import android.view.View;
import android.widget.EditText;
import android.widget.TableLayout;
import android.widget.TextView;
import android.widget.Toast;

public class DashboardActivity extends AppCompatActivity {

    private String url;
    private String userid;
    private String username;

    private TextView welcome_text;
    private EditText query_text;
    private TableLayout table_layout;
    private TextView textView;
    private CountDownTimer timer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dashboard);

        url = getIntent().getStringExtra("url");
        userid = getIntent().getStringExtra("userid");
        username = getIntent().getStringExtra("username");

        welcome_text = findViewById(R.id.welcome_text);
        query_text = findViewById(R.id.query_text);
        table_layout = findViewById(R.id.table_layout);
        textView = findViewById(R.id.textView);

        String welcome_msg = "Welcome, "+username+"!";
        welcome_text.setText(welcome_msg);

        startTimer(this);
    }

    public void onSend(View v) {
        this.timer.cancel();
        startTimer(this);
        String query = query_text.getText().toString();
        if (query.isEmpty()) {
            Toast.makeText(this, "Enter a valid query", Toast.LENGTH_LONG).show();
            return;
        }
        APIHandler apiHandler = new APIHandler(DashboardActivity.this, url, getLayoutInflater(), table_layout);
        apiHandler.sendRequest(userid+"/"+query);
    }

    public void onLogout(View view){
        this.timer.cancel();
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
    }

    public void startTimer(Context context) {
        this.timer = new CountDownTimer(5*60*1000, 1000) {

            public void onTick(long millisUntilFinished) {}

            public void onFinish() {
                Intent intent = new Intent(context, MainActivity.class);
                startActivity(intent);
            }
        };
        this.timer.start();
    }

}

File: android_app/app/src/main/java/com/example/androidapp/DashboardActivity.kt
package com.example.androidapp

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_dashboard.*
import org.json.JSONArray




class DashboardActivity : AppCompatActivity() {
    private lateinit var url: String
    private lateinit var userid: String
    private lateinit var username: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dashboard)

        url = intent.getStringExtra("url").toString()
        userid = intent.getStringExtra("userid").toString()
        username = intent.getStringExtra("username").toString()

        welcome_text.text = getString(R.string.welcome_message, username)

        send_button.setOnClickListener {
            val query = query_text.text.toString()
            if (query.isEmpty()) {
                Toast.makeText(this, "Enter a valid query", Toast.LENGTH_LONG).show()
            }
            APIHandler(this, url).sendRequest("$userid/$query") {
                // TODO
                println(it)
                createTable(it)
            }
        }

        logout_button.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
        }
    }

    private fun createTable(jsonArray: JSONArray) {
        if (jsonArray.length() == 0){
            Toast.makeText(this, "No data matches your query.", Toast.LENGTH_LONG).show()
            return
        }
        table_layout.removeAllViews()
        var row = TableRow(this)
        var jsonObject = jsonArray.getJSONObject(0)
        var iterator: Iterator<String> = jsonObject.keys()
        while (iterator.hasNext()) {
            val key = iterator.next()
            val textview = layoutInflater.inflate(R.layout.headertext, null) as TextView
            textview.text = key
            row.addView(textview)
        }
        table_layout.addView(row)
        for (i in 0 until jsonArray.length()) {
            row = TableRow(this)
            jsonObject = jsonArray.getJSONObject(i)
            iterator = jsonObject.keys()
            while (iterator.hasNext()) {
                val key = iterator.next()
                val textview = layoutInflater.inflate(R.layout.rowtext, null) as TextView
                textview.text = jsonObject.getString(key)
                row.addView(textview)
            }
            table_layout.addView(row)
        }
    }
}
